File: tools/find_missing_translations.js
const fs = require('fs');
const path = require('path');
const basePath = path.join('lib','l10n','app_en.arb');
const locales = ['fr','sw','rw'];

function load(p) {
  return JSON.parse(fs.readFileSync(p,'utf8'));
}

const base = load(basePath);
const baseKeys = Object.keys(base).filter(k => !k.startsWith('@'));
for (let loc of locales) {
  const p = path.join('lib','l10n',`app_${loc}.arb`);
  if (!fs.existsSync(p)) { console.log(`${loc}: file not found`); continue; }
  const obj = load(p);
  const keys = Object.keys(obj).filter(k => !k.startsWith('@'));
  const missing = baseKeys.filter(k => !keys.includes(k));
  console.log(`${loc}: ${missing.length} missing`);
  missing.forEach(k => console.log('  -', k));
}

File: tools/find_missing_translations.py
import json
from pathlib import Path

base = Path('lib/l10n/app_en.arb')
locales = ['fr','sw','rw']

base_keys = set()
with base.open(encoding='utf-8') as f:
    data = json.load(f)
    # ignore metadata keys starting with @
    base_keys = set(k for k in data.keys() if not k.startswith('@'))

missing = {}
for loc in locales:
    p = Path(f'lib/l10n/app_{loc}.arb')
    if not p.exists():
        missing[loc] = None
        continue
    with p.open(encoding='utf-8') as f:
        d = json.load(f)
        keys = set(k for k in d.keys() if not k.startswith('@'))
        missing_keys = sorted(list(base_keys - keys))
        missing[loc] = missing_keys

for loc, keys in missing.items():
    if keys is None:
        print(f'{loc}: file not found')
    else:
        print(f'{loc}: {len(keys)} missing')
        for k in keys:
            print('  -', k)
